In [1]:
import numpy as np

This is a very simple version of K means clustering using a 1-Dimensional array.

In [2]:
data = np.array([23,34,43,26,41,78,86,68,97,76,150,135,160,125,144])

 We choose the initial random centroids from the data array.

In [3]:
def ini_choose_centroids(data, k):
    x,y,z = np.random.choice(data, size=k, replace=False)
    return x,y,z

print(ini_choose_centroids(data))



(23, 76, 26)


Using these centroids, we find the lowest absolute differences from points in the data to the centroids. We then append the corresponding points to the corresponding cluster. If the centroids that have been chosen leave a cluster empty, we randomise them again and repeat the process so that all clusters have at least 1 value to provide a mean. 

In [4]:

def cluster_values(centroid1, centroid2, centroid3):
    cluster1 = []
    cluster2 = []
    cluster3 = []
    #print(ini_choose_centroids(data)[0])
    for i in range(len(data)):
        if abs(centroid1 - data[i]) < abs(centroid2 - data[i]) and abs(centroid1 - data[i]) < abs(centroid3 - data[i]):
            cluster1.append(data[i])
        elif abs(centroid2 - data[i]) < abs(centroid1 - data[i]) and abs(centroid2 - data[i]) < abs(centroid3 - data[i]):
            cluster2.append(data[i])
        else:
            cluster3.append(data[i])
    if len(cluster1) == 0 or len(cluster2) == 0 or len(cluster3) == 0:
        return cluster_values(ini_choose_centroids(data)[0], ini_choose_centroids(data)[1], ini_choose_centroids(data)[2])
    return cluster1, cluster2, cluster3
print(cluster_values(ini_choose_centroids(data)[0], ini_choose_centroids(data)[1], ini_choose_centroids(data)[2]))

([34], [43, 41, 78, 86, 68, 97, 76, 150, 135, 160, 125, 144], [23, 26])


We use this to calculate and then update the cluster means. Once the means have been calculated, they are fed back into the cluster_values function and act as the new centroids. 

In [5]:
def update(cluster_1, cluster_2, cluster_3, ):
    clust_mean1 = np.mean(cluster_1)
    clust_mean2 = np.mean(cluster_2)
    clust_mean3 = np.mean(cluster_3)
    print('mean = ' + str(clust_mean1))
    print('mean = ' + str(clust_mean2))
    print('mean = ' + str(clust_mean3))
    return cluster_values(clust_mean1, clust_mean2, clust_mean3)

print(update([23, 34, 43, 26, 41], [78, 86, 68, 76], [97, 150, 135, 160, 125, 144]))

mean = 33.4
mean = 77.0
mean = 135.16666666666666
([23, 34, 43, 26, 41], [78, 86, 68, 97, 76], [150, 135, 160, 125, 144])


Here is the full function. It chooses the centroids and creates the initial clusters. It then finds and updates the means of the clusters and feeds them back to the clustering function. Then, within a loop that runs 9 times (for a total of 10), it repeats this process and hopefully by the end the data means do not move and the optimal clustering has been achieved. 

In [6]:
def k_means():
    centroids = ini_choose_centroids(data)
    clusters = cluster_values(centroids[0], centroids[1], centroids[2])
    updates = update(clusters[0],clusters[1],clusters[2])
    for i in range(9):
        print(updates)
        updates = update(updates[0], updates[1], updates[2])
    return updates
print(k_means())

mean = 102.66666666666667
mean = 48.625
mean = 147.25
([78, 86, 97, 76], [23, 34, 43, 26, 41, 68], [150, 135, 160, 125, 144])
mean = 84.25
mean = 39.166666666666664
mean = 142.8
([78, 86, 68, 97, 76], [23, 34, 43, 26, 41], [150, 135, 160, 125, 144])
mean = 81.0
mean = 33.4
mean = 142.8
([78, 86, 68, 97, 76], [23, 34, 43, 26, 41], [150, 135, 160, 125, 144])
mean = 81.0
mean = 33.4
mean = 142.8
([78, 86, 68, 97, 76], [23, 34, 43, 26, 41], [150, 135, 160, 125, 144])
mean = 81.0
mean = 33.4
mean = 142.8
([78, 86, 68, 97, 76], [23, 34, 43, 26, 41], [150, 135, 160, 125, 144])
mean = 81.0
mean = 33.4
mean = 142.8
([78, 86, 68, 97, 76], [23, 34, 43, 26, 41], [150, 135, 160, 125, 144])
mean = 81.0
mean = 33.4
mean = 142.8
([78, 86, 68, 97, 76], [23, 34, 43, 26, 41], [150, 135, 160, 125, 144])
mean = 81.0
mean = 33.4
mean = 142.8
([78, 86, 68, 97, 76], [23, 34, 43, 26, 41], [150, 135, 160, 125, 144])
mean = 81.0
mean = 33.4
mean = 142.8
([78, 86, 68, 97, 76], [23, 34, 43, 26, 41], [150, 135, 160

The optimal means appear to be 142.8, 33.4 and 81.0 with clusters of [150, 135, 160, 125, 144], [23, 34, 43, 26, 41] and [78, 86, 68, 97, 76]